### PR TITLE
Fixed iterator regression: Avoid using heap for each sample when iterating.

### PR DESCRIPTION
**Pull request by bwplotka**
Fixes: https://github.com/prometheus/prometheus/issues/7873

Signed-off-by: Bartlomiej Plotka <bwplotka@gmail.com>

### CODE REVIEW

**Comment by bwplotka** [comment]
Previously vs prior change:
```
name                                   old time/op    new time/op    delta
DBBlocksQuerierSelect/[__name__="up"]    53.0ms ± 0%   489.5ms ± 0%  +823.32%

name                                   old alloc/op   new alloc/op   delta
DBBlocksQuerierSelect/[__name__="up"]    1.86MB ± 0%    3.71MB ± 0%   +98.80%

name                                   old allocs/op  new allocs/op  delta
DBBlocksQuerierSelect/[__name__="up"]     49.5k ± 0%     67.9k ± 0%   +37.18%
```

Now vs prior change:
```
name                                   old time/op    new time/op    delta
DBBlocksQuerierSelect/[__name__="up"]    52.7ms ± 0%    81.1ms ± 0%  +53.97%

name                                   old alloc/op   new alloc/op   delta
DBBlocksQuerierSelect/[__name__="up"]    1.86MB ± 0%    1.96MB ± 0%   +5.18%

name                                   old allocs/op  new allocs/op  delta
DBBlocksQuerierSelect/[__name__="up"]     49.5k ± 0%     52.4k ± 0%   +5.80%
```

**Comment by brian-brazil** [comment]
It is probably simpler to use Fix rather than doing all the curr bookkeeping yourself.

File: storage/merge.go
Comment (bwplotka) [reply storage/merge.go:481]: True, but we don't care that much, it's just nice to have it set to something so we don't need to care to do if curr != nil

File: storage/merge.go
Comment (bwplotka) [reply storage/merge.go]: good catch,

**Comment by bwplotka** [comment]
>  use Fix ?

File: storage/merge.go
Comment (bwplotka) [inline storage/merge.go:493]: All should work, but I am not huge fan of checking duplicates within a chunk/single iterator itself :thinking: 

Let me benchmark it if it has any overhead...

Comment (bwplotka) [reply storage/merge.go:493]: No overhead.

**Review by brian-brazil — COMMENTED** [review]
> Sorry missing this, what do you mean?

The heap.Fix method. Right now you're having to do a lot of potentially fragile bookkeeping with curr, which using that approach would avoid as there would only be the heap.

**Comment by bwplotka** [comment]
Wow I did not see the fix method before. Not sure what it does but if it helps with bookkeeping, I would definitely check it out and use it.

I am on holidays this week until Friday, so I would suggest merging this as it fixes regression, so we unblock the release. Then when I am back I can adjust it to use Fix instead, look for further perf improvements and properly detect failure cases as comments suggests 🤙

cc @brian-brazil @roidelapluie

**Comment by roidelapluie** [comment]
@codesome @krasi-georgiev

What do you think about this pull request ? Is it good for merging into release-2.21? If not, can you improve it in the branch  (Bartek is in holidays this week)

**Comment by codesome** [comment]
Will check shortly. We need to fix https://github.com/prometheus/prometheus/pull/7900#discussion_r483798322 anyway.

File: storage/merge.go
Comment (codesome) [reply storage/merge.go:473]: @brian-brazil it should be fixed now

**Review by codesome — APPROVED** [review]
LGTM from me

File: storage/merge.go
Comment (brian-brazil) [reply storage/merge.go:473]: Couldldn't we just check c.curr?

Comment (codesome) [reply storage/merge.go:473]: _Right_

File: storage/merge.go
Comment (brian-brazil) [inline storage/merge.go:495]: Don't we need to push&pop/fix here so that c.curr is updated?

File: storage/merge.go
Comment (brian-brazil) [inline storage/merge.go:513]: If we've gotten here, isn't this always going to be true?

File: storage/merge.go
Comment (codesome) [reply storage/merge.go:495]: Not required. It will follow the same loop and compare the next in the curr with the top of the heap and do the push and pop if required.

File: storage/merge.go
Comment (codesome) [reply storage/merge.go:513]: We come here when either curr is exhausted or it has crossed the top of the heap. It is possible that the top of the heap has the same timestamp as the lastt.

File: storage/merge.go
Comment (brian-brazil) [inline storage/merge.go:469]: c.curr should be set to nil here

**Review by brian-brazil — COMMENTED** [review]
I've been staring at this code for the last hour, and I'm still not 100% confident that it is correct. The cyclomatic complexity is high, with non-obvious dependencies, invariants you think would be present but aren't, and many subtleties.

At a minimum this needs comments explaining how this is meant to work, and the state at each point.

A simpler implementation would be preferable, this looks to be unmaintainable - and I suspect our unittests don't cover all of the corner cases that this code now adds.

File: storage/merge.go
Comment (codesome) [inline storage/merge.go:498]: (Review for myself) we are not setting `c.lastt` here.

**Review by codesome — COMMENTED** [review]
I am trying to simplify it now (while that will make it more verbose)

**Comment by codesome** [comment]
I have updated the PR with more comments and tiny refactoring

File: storage/merge.go
Comment (codesome) [inline storage/merge.go:524]: Hm, this is wrong if c.curr.Next() returned false. I guess the previous one was more apt.

File: storage/merge.go
Comment (brian-brazil) [inline storage/merge.go:508]: `no`

**Comment by roidelapluie** [comment]
CI is failing:

```
storage/merge.go:499:3: ineffectual assignment to `currt` (ineffassign)
		currt = math.MaxInt64
```

File: storage/merge.go
Comment (brian-brazil) [inline storage/merge.go:499]: This value seems to be never used.

Comment (codesome) [reply storage/merge.go:499]: RIght, my bad

**Comment by brian-brazil** [comment]
:+1:

**Comment by codesome** [comment]
I will let @roidelapluie merge this as he is 2.21 release shepherd.

**Comment by bwplotka** [comment]
I'm back, thanks for shepherding this @codesome :+1:  Good work!

I will look later on if we can replace this with https://golang.org/src/container/heap/heap.go?s=2875:2903#L74

**Pull request merged**
